br/pkg/storage/mapstore.go

```diff
@@ -0,0 +1,304 @@
+	"go.uber.org/atomic"
+)
+
+type mapFile struct {
+	sync.RWMutex
+	Data []byte
+}
+
+type mapStorage struct {
+	sync.RWMutex
+	dataStore map[string]*mapFile
+}
+
+func NewMapStorage() *mapStorage {
+	return &mapStorage{
+		dataStore: make(map[string]*mapFile),
+	}
+}
+
+// DeleteFile delete the file in storage
+func (s *mapStorage) DeleteFile(ctx context.Context, name string) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		// continue on
+	}
+	if !path.IsAbs(name) {
+		return errors.Errorf("file name is not an absolute path: %s", name)
+	}
+	s.Lock()
+	defer s.Unlock()
+	if _, ok := s.dataStore[name]; !ok {
+		return errors.Errorf("cannot find the file: %s", name)
+	}
+	delete(s.dataStore, name)
+	return nil
+}
+
+// WriteFile file to storage.
+func (s *mapStorage) WriteFile(ctx context.Context, name string, data []byte) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		// continue on
+	}
+	if !path.IsAbs(name) {
+		return errors.Errorf("file name is not an absolute path: %s", name)
+	}
+	r := bytes.NewReader(data)
+	fileData, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	s.Lock()
+	defer s.Unlock()
+	theFile, ok := s.dataStore[name]
+	if ok {
+		theFile.Lock()
+		theFile.Data = fileData
+		theFile.Unlock()
+	} else {
+		s.dataStore[name] = &mapFile{
+			Data: fileData,
+		}
+	}
+	return nil
+}
+
+// ReadFile storage file.
+func (s *mapStorage) ReadFile(ctx context.Context, name string) ([]byte, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+		// continue on
+	}
+	if !path.IsAbs(name) {
+		return nil, errors.Errorf("file name is not an absolute path: %s", name)
+	}
+	s.RLock()
+	defer s.RUnlock()
+	theFile, ok := s.dataStore[name]
+	if !ok {
+		return nil, errors.Errorf("cannot find the file: %s", name)
+	}
+	theFile.RLock()
+	defer theFile.RUnlock()
+	r := bytes.NewReader(theFile.Data)
+	retData, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	return retData, nil
+}
+
+// FileExists return true if file exists.
+func (s *mapStorage) FileExists(ctx context.Context, name string) (bool, error) {
+	select {
+	case <-ctx.Done():
+		return false, ctx.Err()
+	default:
+		// continue on
+	}
+	if !path.IsAbs(name) {
+		return false, errors.Errorf("file name is not an absolute path: %s", name)
+	}
+	s.RLock()
+	defer s.RUnlock()
+	_, ok := s.dataStore[name]
+	return ok, nil
+}
+
+// Open a Reader by file path.
+func (s *mapStorage) Open(ctx context.Context, filePath string) (ExternalFileReader, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+		// continue on
+	}
+	if !path.IsAbs(filePath) {
+		return nil, errors.Errorf("file name is not an absolute path: %s", filePath)
+	}
+	s.RLock()
+	defer s.RUnlock()
+	theFile, ok := s.dataStore[filePath]
+	if !ok {
+		return nil, errors.Errorf("cannot find the file: %s", filePath)
+	}
+	theFile.RLock()
+	defer theFile.RUnlock()
+	r := bytes.NewReader(theFile.Data)
```

Comment:
This won't happen, because the implementation of write operations of the `mapstore` is 'buffer replacement' rather than 'in-place buffer modification'.  This means after the `bytes.Reader` of `theFile.Data` is generated, even if the `theFile.Data` is replaced in the future, the reader can always read the content of the original buffer without errors.  The RW-lock here is to ensure that the buffer is not changed when the `bytes.Reader` is being created.

Reply:
Perhaps you are right. But this confuses me that the file lock doesn't need to hold when reading its content. Could you add some comments or choose a more proper name for the lock, which just guards the data pointer rather than the data content? A simple way is to use `atomic.Value`.